[M] ENT-5194: Migrate rest of the owner_resource_spec to Java

bakajstep opened the pull request:
- Migrate owner_resource_spec.rb
- Delete ownere_resource_spec.rb
- Create enum for user permissions

Ceiu commented on spec-tests/src/test/java/org/candlepin/spec/owners/OwnerResourcePoolFilterSpecTest.java:
```diff
@@ -0,0 +1,322 @@
+            ProductAttributes.VirtLimit.withValue("unlimited"));
+        PoolDTO pool;
+
+        if (CandlepinMode.isHosted()) {
```
Unnecessary hosted check. Just create the pools as you need them.

bakajstep replied:
done

Ceiu reviewed (changes requested):
A couple overarching questions:
- What drove the decision to put these tests in separate files entirely?
- Do any of the tests making use of the call to `isHosted` actually care about Candlepin's operating mode? Is refresh a critical component of the test? Is there an explicit reason to differentiate where the subscription information originates?

bakajstep commented on api/candlepin-api-spec.yaml:
```diff
@@ -4767,7 +4767,8 @@ paths:
           schema:
             type: array
             items:
-              $ref: '#/components/schemas/KeyValueParamDTO'
```
We need to change this because, it wasn't working. Our client parse that object into string and send it in request in wrong way. Will already face this issue before me and came with this solution. I just "stole" his solution in his PR.

bakajstep commented:
> A couple overarching questions:
> 
> * What drove the decision to put these tests in separate files entirely?
> * Do any of the tests making use of the call to `isHosted` actually care about Candlepin's operating mode? Is refresh a critical component of the test? Is there an explicit reason to differentiate where the subscription information originates?

I split them because it was many classes in one file so instead we would have one big file i split them into smaller ones.

Januson commented on spec-tests/src/test/java/org/candlepin/spec/owners/OwnerResourceSpecTest.java:
```diff
@@ -317,6 +322,18 @@ void onlySuperAdminCanRefresh() throws ApiException {
 
+        ProductDTO product = Products.random();
+
+        if (CandlepinMode.isHosted()) {
```
Why do names have to match? The point of display name annotation is to allow nice sentence-like name without the need to use such a construct an actual test name. If we are to duplicate the name we could just skip the @DisplayName annotation completely and use @DisplayNameGenerator instead to convert camel-case to spaces.

Ceiu replied:
Because otherwise when we have a failing test, I have to do a find-string-in-project/folder to find the offending test to debug and fix it, whereas if I can see the name, I can jump to the symbol. This is even more of a thing as our test output keeps getting worse and worse.

bakajstep commented on spec-tests/src/test/java/org/candlepin/spec/owners/OwnerResourcePoolFilterSpecTest.java:
```diff
@@ -0,0 +1,306 @@
+        @Test
+        @DisplayName("should allow owners filter pools by single filter")
+        public void shouldAllowOwnersFilterPoolsBySingleFilter() throws ApiException {
+            List<PoolDTO> pools = owners.listOwnerPools(owner.getKey());
```
i forgot there the line, because i was rewriting it to nested test suite.

bakajstep replied:
done

bakajstep commented on spec-tests/src/test/java/org/candlepin/spec/owners/OwnerResourceSpecTest.java:
```diff
@@ -731,6 +748,80 @@ void listSystemPurposeAttributes() throws ApiException {
+
+    @Test
+    @DisplayName("should allow user with owner pools permission can see system purpose of the owner products")
+    void shouldAllowUserWithOwnerPoolsPermissionCanSeeSystemPurposeOfTheOwnerProducts() throws ApiException {
```
done

bakajstep commented on spec-tests/src/test/java/org/candlepin/spec/owners/OwnerResourcePoolFilterSpecTest.java:
```diff
@@ -0,0 +1,306 @@
+    @Nested
+    @DisplayName("Pool and subscription filtering")
+    @TestInstance(TestInstance.Lifecycle.PER_CLASS)
+    class PoolAndSubscriptionFiltering {
```
done

bakajstep commented on spec-tests/src/test/java/org/candlepin/spec/owners/OwnerResourcePoolFilterSpecTest.java:
```diff
@@ -0,0 +1,306 @@
+            );
+            product3 = ownerProducts.createProductByOwner(owner.getKey(), Products.random());
+
+            if (CandlepinMode.isHosted()) {
```
done

bakajstep commented on spec-tests/src/main/java/org/candlepin/spec/bootstrap/data/builder/Roles.java:
```diff
@@ -20,30 +20,29 @@
-            .name(StringUtil.random("test-role-", 8, StringUtil.CHARSET_NUMERIC_HEX))
-            .permissions(List.of(ownerPermission));
+            .name(StringUtil.random("test-role"))
+            .permissions(Arrays.asList(permissions));
```
ok, done

bakajstep commented on spec-tests/src/test/java/org/candlepin/spec/owners/OwnerResourceFuturePoolSpecTest.java:
```diff
@@ -0,0 +1,265 @@
+
+@SpecTest
+public class OwnerResourceFuturePoolSpecTest {
+
```
if you meaning between test and variables i change it to 1 empty line

bakajstep commented on spec-tests/src/test/java/org/candlepin/spec/owners/OwnerResourceFuturePoolSpecTest.java:
```diff
@@ -0,0 +1,265 @@
+        private PoolDTO futurePool2;
+
+        @BeforeAll
+        void setup() throws ApiException {
```
done

bakajstep commented on spec-tests/src/test/java/org/candlepin/spec/owners/OwnerResourceSpecTest.java:
```diff
@@ -731,6 +748,80 @@ void listSystemPurposeAttributes() throws ApiException {
+        List<RoleDTO> roles = userClient.users().getUserRoles(user.getUsername());
+        assertEquals(1, roles.size());
+
+        //The user should have access to the owner's system purpose attributes (expecting this will not return
```
done

bakajstep commented on src/test/java/org/candlepin/resource/OwnerResourceTest.java:
```diff
@@ -980,19 +980,13 @@ public void testListConsumersByHypervisorId() {
-            this.buildFactParam("fact-1", "value-1b"),
-            this.buildFactParam("fact-2", "value-2"),
-            this.buildFactParam("fact-3", "value-3"));
+        List<String> factsParam = Arrays.asList(
```
done

bakajstep commented on src/main/java/org/candlepin/resource/OwnerResource.java:
```diff
@@ -1238,8 +1238,14 @@ public Stream<PoolDTO> listOwnerPools(
+                poolFilters.addAttributeFilter(keyValue[0], keyValue[1]);
+            }
+            else {
+                throw new BadRequestException("Illegal filter argument format!");
```
yes, you are right i throw wrong exception
https://github.com/candlepin/candlepin/blob/0210260b8ad1ba7f0f282eea3d5b2b18ece73c00/src/main/java/org/candlepin/resteasy/converter/KeyValueParamConverterProvider.java#L52-L66

bakajstep replied:
fixed

bakajstep replied:
and it should be same as previous

Ceiu commented on src/main/java/org/candlepin/resource/OwnerResource.java:
```diff
@@ -1239,8 +1240,14 @@ public Stream<PoolDTO> listOwnerPools(
-            poolFilters.addAttributeFilter(filterParam.getKey(), filterParam.getValue());
+        for (String filterParam : attrFilters) {
+            String[] keyValue = filterParam.split(":");
+            if (keyValue.length == 2) {
```
This check doesn't line up with how we perform this processing in other places, even within the same file, nor does it match with the previous behavior. Previous builds split the key:value on the first instance of a colon, with all others being part of the value.

Whenever we make API changes like this, we need to ensure that Katello and Teamnado, at least, will not be negatively affected by such a change. Otherwise we need to keep the behavior the same. Additionally, we want to be consistent in how we accept these things, as it's very disruptive and confusing to users if a format is acceptable in one place, not in another, behaves differently in a third, and so on.

Ceiu replied:
Right, the main difference here is the original way of parsing it left colons in the value alone. That is, the string `"key:value:with:colons"` would be parsed to:
- key: "key"
- value: "value:with:colons"

The parsing code in this PR has a few different methods. One will split the string into "key" and "value" and then silently discard the rest. Others will split into those two strings and then throw an error because of the extra colons. Neither of which lines up with the old behavior of only splitting on the first colon and then only throwing an exception if there was nothing to split on at all.

In the name of standardizing all of this: do we have a way to use the old converter again, such that each time we receive a key-value parameter we aren't manually parsing it ourselves? If not, then let's add a utility function for doing this parsing one or more key-value pairs so we have a centralized place that is guaranteed to be following the same rules.

As far as the translation goes, I was surprised by the CandlepinParameterParseException because that is *also* not at all how we're supposed to generate those, but since that's how we've done it in the past, that's acceptable here.

bakajstep replied:
i create class with old parsing so it should be working now.

bakajstep commented on src/main/java/org/candlepin/resource/OwnerResource.java:
```diff
@@ -1239,8 +1240,14 @@ public Stream<PoolDTO> listOwnerPools(
+                poolFilters.addAttributeFilter(keyValue[0], keyValue[1]);
+            }
+            else {
+                throw new CandlepinParameterParseException("name:value");
```
it does because you have mapper for it to map it to 400. i add another param so user now know which parsed attributed is failing. 
"""
Runtime Error failed to parsed: support_level;Supurb, expected format: key:value at org.candlepin.resource.OwnerResource.listOwnerPools:1,249
"""
and here you can see the translation
https://github.com/candlepin/candlepin/blob/003e21e8deb95c4a0f976d5a5e8faf698f9e8dfa/src/main/java/org/candlepin/exceptions/mappers/CandlepinExceptionMapper.java#L103-L107

bakajstep commented on spec-tests/src/test/java/org/candlepin/spec/owners/OwnerResourceFuturePoolSpecTest.java:
```diff
@@ -0,0 +1,265 @@
+public class OwnerResourceFuturePoolSpecTest {
+
+
+    private OffsetDateTime now;
```
not necessary, but we are using it in most of the tests. I replace it with constants it look better and code is more readable.

bakajstep commented on spec-tests/src/test/java/org/candlepin/spec/owners/OwnerResourceFuturePoolSpecTest.java:
```diff
@@ -0,0 +1,265 @@
+
+        ProductDTO product = createProductByOwner(adminClient, owner);
+        PoolDTO futurePool;
+        if (CandlepinMode.isHosted()) {
```
done

bakajstep commented on spec-tests/src/test/java/org/candlepin/spec/owners/OwnerResourceFuturePoolSpecTest.java:
```diff
@@ -0,0 +1,265 @@
+            ProductDTO product1 = createProductByOwner(adminClient, owner);
+            ProductDTO product2 = createProductByOwner(adminClient, owner);
+
+            if (CandlepinMode.isHosted()) {
```
done

bakajstep commented on spec-tests/src/test/java/org/candlepin/spec/owners/OwnerResourceOwnerInfoSpecTest.java:
```diff
@@ -0,0 +1,161 @@
+            ProductAttributes.MultiEntitlement.withValue("true")
+        );
+        ProductDTO product = adminClient.ownerProducts().createProductByOwner(owner.getKey(), tmpProduct);
+        if (CandlepinMode.isHosted()) {
```
done

bakajstep commented on spec-tests/src/test/java/org/candlepin/spec/owners/OwnerResourcePoolFilterSpecTest.java:
```diff
@@ -0,0 +1,297 @@
+        ProductDTO product = createProductByOwner(adminClient, owner, Products.random());
+        PoolDTO pool;
+        // entitle owner for the virtual and monitoring products
+        if (CandlepinMode.isHosted()) {
```
done

bakajstep commented on src/main/java/org/candlepin/resource/util/EntitlementFinderUtil.java:
```diff
@@ -34,6 +35,9 @@ public static EntitlementFilterBuilder createFilter(
             for (String filterParam : attrFilters) {
                 String[] keyValue = filterParam.split(":");
+                if (keyValue.length != 2) {
+                    throw new BadRequestException("Illegal attribute format!");
```
i replaced it with CandlepinParameterParseException. I forgot about this.

Ceiu reviewed (changes requested):
I'm still not sure where I stand in regards to the many files thing, but it's probably fine for now until the standards doc is done and accepted.

Ceiu commented on src/main/java/org/candlepin/exceptions/CandlepinParameterParseException.java:
```diff
@@ -22,9 +22,10 @@ public class CandlepinParameterParseException extends RuntimeException {
+    public CandlepinParameterParseException(String expectedFormat, String failedAttribute) {
         // NOTE: Exception message will be generated by BadRequestExceptionMapper.
         //       See the mapper class for details.
+        super(String.format("failed to parsed: %s, expected format: %s", failedAttribute, expectedFormat));
```
This string won't be translated

bakajstep replied:
done

bakajstep commented:
i also remove displayname how we agreed

Ceiu reviewed (changes requested):
In general I'm alright with the concept of a converter -- that's what we had initially, after all -- but it feels like we're coming full circle except manually building and using them ourselves. I feel like if we're going to try to build a set of objects and scaffolding around the conversion, we should work it into the OpenAPI bits somehow to not need to do it manually.

Also, given how much extra work had to be done passing around i18n providers and instances, we should use that time to, instead, cleanup the exception message in the invalid parameter case. The way it worked in the past wasn't exactly great, and standardizing it (and possibly removing that exception entirely) would probably be best off.

We're approaching the point with this that it should probably be its own separate PR, since the rest of the testing seems fine, and is being held up on the fix for the kv param issue.

bakajstep commented on src/main/java/org/candlepin/model/ConsumerCurator.java:
```diff
@@ -173,6 +174,10 @@ public ConsumerQueryArguments addFact(String fact, String value) {
             return this;
         }
 
+        public void addFact(KeyValueParameter keyValueParameter) {
```
done

bakajstep commented on src/main/java/org/candlepin/model/FilterBuilder.java:
```diff
@@ -86,6 +88,10 @@ public void addAttributeFilter(String attrName, String attrValue) {
         attributeFilters.get(attrName).add(attrValue);
     }
 
+    public void addAttributeFilter(KeyValueParameter keyValueParameter) {
```
done

bakajstep commented on src/main/java/org/candlepin/model/dto/KeyValueParameter.java:
```diff
@@ -0,0 +1,58 @@
+     *                            in the format: key:value
+     */
+    public KeyValueParameter(I18n i18n, String inValue) {
+        String[] parts = inValue.split(":", 2); // Maximum of two parts
```
done

bakajstep commented on src/main/java/org/candlepin/model/dto/KeyValueParameter.java:
```diff
@@ -0,0 +1,58 @@
+        }
+
+        key = parts[0];
+        value = parts.length > 1 ? parts[1] : "";
```
done

nikosmoum commented on spec-tests/src/test/java/org/candlepin/spec/owners/OwnerResourceConsumerFactFilterSpecTest.java:
```diff
@@ -0,0 +1,105 @@
+        adminClient = ApiClients.admin();
+        owner = adminClient.owners().createOwner(Owners.random());
+        UserDTO user = UserUtil.createUser(adminClient, owner);
+        ApiClient userClient = ApiClients.trustedUser(user.getUsername());
```
Can we replace all `ApiClients.trustedUser(username)` with `ApiClients.basic(username, password)` through out this PR (and for all other spec tests from now on)? The ApiClients.trustedUser and ApiClients.trustedConsumer should never be used in spec tests (aside from the few tests that explicitly test those authentication methods, e.g. TrustedConsumerAuthSpecTest & TrustedUserAuthSpecTest).

bakajstep replied:
ok, done

Ceiu commented:
The key-value changes (#3623) have been merged, so this PR just needs a rebase around it.

bakajstep commented:
> The key-value changes (#3623) have been merged, so this PR just needs a rebase around it.

done

Ceiu commented on spec-tests/src/test/java/org/candlepin/spec/owners/OwnerResourceEntitlementListSpecTest.java:
```diff
@@ -0,0 +1,104 @@
+            null, null, null, null);
+        assertEquals(1, ents.size());
+
+        List<AttributeDTO> attributes = ents.stream().findFirst()
```
This is an interesting way of doing this, but given that we're testing the whole chain all the way down, this can be done with an assertion chain instead:

```java
assertThat(ents)
  .isNotNull()
  .singleElement()
  .extracting(EntitlementDTO::getPool)
  .isNotNull()
  .extracting(PoolDTO::getProductAttributes, as(collection(AttributeDTO.class)))
  .isNotNull()
  .singleElement()
  .returns("variant", AttributeDTO::getName)
  .returns("Satellite Starter Pack", AttributeDTO::getValue);
```

That way we're verifying the output of our target method down the whole tree, and any issue or incorrect output will be verified, or throw an assertion error rather than an unasserted exception.

bakajstep replied:
done

Ceiu replied:
The old attributes checks still look to be present below them. With this chain in place, the other checks are no longer necessary.

bakajstep replied:
done

Ceiu commented on spec-tests/src/test/java/org/candlepin/spec/owners/OwnerResourceOwnerInfoSpecTest.java:
```diff
@@ -0,0 +1,135 @@
+        assertEquals(5, info.getConsumerCounts().get("system").intValue());
+
+        OwnerInfo systemOwnerInfo = userClientWithPerms.owners().getOwnerInfo(owner.getKey());
+        assertEquals(3, systemOwnerInfo.getConsumerCounts().get("system").intValue());
```
These would be easier to read as map assertions:

```
assertThat(systemOwnerInfo.getConsumerCounts())
  .isNotNull()
  .containsEntry("system", 3);

assertThat(systemOwnerInfo.getConsumerGuestCounts())
  .isNotNull()
  .containsEntry("physical", 1)
  .containsEntry("guest", 2);

...
```

bakajstep replied:
done

Ceiu commented on spec-tests/src/test/java/org/candlepin/spec/owners/OwnerResourcePoolFilterSpecTest.java:
```diff
@@ -0,0 +1,254 @@
+
+        @Test
+        public void shouldAllowOwnersFilterPoolsByMultiplePoolId() throws ApiException {
+            List<PoolDTO> pools = adminClient.owners().listOwnerPools(owner.getKey());
```
We should add an assertion on this to verify that this output contains at least two elements, as the remainder of the test is reliant upon that.

bakajstep replied:
done

Ceiu commented on spec-tests/src/test/java/org/candlepin/spec/owners/OwnerResourcePoolFilterSpecTest.java:
```diff
@@ -0,0 +1,254 @@
+            null, null, null, null, List.of(providedName), null,
+            null, null, null, null, null, null, null, null);
+        assertEquals(1, pools.size());
+        assertEquals(owner.getKey(), pools.get(0).getOwner().getKey());
```
Since we're targeting stuff off the first pool of the list, we could extract it and clean up the repetition and UuT fetching a bit:

```java

assertThat(pools)
  .isNotNull()
  .hasSize(1);

PoolDTO pool = pools.get(0);
assertThat(pool)
  .isNotNull()
  .returns(product.getId(), PoolDTO::getProductId)
  .extracting(PoolDTO::getProvidedProducts)
  .isNotNull()
  .hasSize(3);

assertThat(pool.getOwner())
  .isNotNull()
  .returns(owner.getKey(), OwnerDTO::getKey);
```

bakajstep replied:
done

Ceiu commented on spec-tests/src/test/java/org/candlepin/spec/owners/OwnerResourcePoolFilterSpecTest.java:
```diff
@@ -0,0 +1,254 @@
+        public void shouldAllowOwnersFilterPoolsBySingleFilter() throws ApiException {
+            List<PoolDTO> filteredPools = adminClient.owners().listOwnerPoolsWithAttributes(
+                owner.getKey(), List.of("support_level:VIP"));
+            assertEquals(1, filteredPools.size());
```
These could be combined into one assertion chain that includes the missing null checks:

```java
assertThat(filteredPools)
  .isNotNull()
  .singleElement()
  .isNotNull()
  .returns(product1.getId(), PoolDTO::getProductId);
```

This could be applied in several places.

bakajstep replied:
done, I replaced it in other files where it seemed appropriate, so I hope I didn't forget it somewhere

Ceiu reviewed (changes requested):
Looking pretty good. Just a few cleanup things and more usage of assertJ; specifically to get those null checks in there.
